compiler/src/dotty/tools/dotc/Compiler.scala

```diff
@@ -40,77 +40,92 @@ class Compiler {
-           new ArrayConstructors),  // Intercept creation of (non-generic) arrays and intrinsify.
-      List(new Erasure),            // Rewrite types to JVM model, erasing all type parameters, abstract types and refinements.
-      List(new ElimErasedValueType, // Expand erased value types to their underlying implmementation types
-           new VCElideAllocations,  // Peep-hole optimization to eliminate unnecessary value class allocations
-           new Mixin,               // Expand trait fields and trait initializers
-           new LazyVals,            // Expand lazy vals
-           new Memoize,             // Add private fields to getters and setters
-           new NonLocalReturns,     // Expand non-local returns
-           new CapturedVars),       // Represent vars captured by closures as heap objects
-      List(new Constructors,        // Collect initialization code in primary constructors
-                                    // Note: constructors changes decls in transformTemplate, no InfoTransformers should be added after it
-           new FunctionalInterfaces, // Rewrites closures to implement @specialized types of Functions.
-           new GetClass,            // Rewrites getClass calls on primitive types.
-           new Simplify),           // Perform local optimizations, simplified versions of what linker does.
-      List(new LinkScala2Impls,     // Redirect calls to trait methods defined by Scala 2.x, so that they now go to their implementations
-           new LambdaLift,          // Lifts out nested functions to class scope, storing free variables in environments
+    frontendPhases ::: picklerPhases ::: transformPhases ::: backendPhases
+
+  /** Phases dealing with the the frontend up to trees ready for TASTY pickling */
+  protected def frontendPhases: List[List[Phase]] =
+    List(new FrontEnd) ::          // Compiler frontend: scanner, parser, namer, typer
+    List(new sbt.ExtractDependencies) :: // Sends information on classes' dependencies to sbt via callbacks
+    List(new PostTyper) ::         // Additional checks and cleanups after type checking
+    List(new sbt.ExtractAPI) ::    // Sends a representation of the API of classes to sbt via callbacks
+    Nil
+
+  /** Phases dealing with TASTY tree pickling and unpickling */
+  protected def picklerPhases: List[List[Phase]] =
+    List(new Pickler) ::           // Generate TASTY info
+    List(new LinkAll) ::           // Reload compilation units from TASTY for library code (if needed)
+    Nil
+
+  /** Phases dealing transforming from pickled trees to backend trees */
+  protected def transformPhases: List[List[Phase]] =
+    List(new FirstTransform,        // Some transformations to put trees into a canonical form
+         new CheckReentrant,        // Internal use only: Check that compiled program has no data races involving global vars
+         new ElimJavaPackages) ::   // Eliminate syntactic references to Java packages
```

Comment:
Why use `::` instead of `List(List(...), ...)` ?

Reply:
I personally prefer `List(List(...), ...)` 😄

Reply:
I thought it would make a better distinction between micro and macro phases and it is a little bit more efficient.